File: go_algorithms/algorithms/WinogradScaled.go
package algorithms

import (
	"fmt"
	"math"
	"time"
)

func WinogradScaled(A, B [][]int) ([][]int, int64) {

	fmt.Println("Método: WinogradScaled")

	startTime := time.Now().UnixNano()

	N := len(B[0])
	P := len(B)
	M := len(A)

	CopyA := make([][]int, N)
	for i := range CopyA {
		CopyA[i] = make([]int, P)
		copy(CopyA[i], A[i])
	}

	CopyB := make([][]int, P)
	for i := range CopyB {
		CopyB[i] = make([]int, M)
		copy(CopyB[i], B[i])
	}

	// Scaling factors
	a := NormInf(A, N, P)
	b := NormInf(B, P, M)
	lambda := math.Floor(0.5 + math.Log(float64(b)/float64(a))/math.Log(4))

	// Scaling
	MultiplyWithScalar(A, CopyA, N, P, math.Pow(2, float64(lambda)))
	MultiplyWithScalar(B, CopyB, P, M, math.Pow(2, -float64(lambda)))

	// Using Winograd with scaled matrices
	result := Winograd(CopyA, CopyB)

	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime
	return result, executionTime
}

func NormInf(matrix [][]int, rows, cols int) int {
	maxNorm := 0
	for i := 0; i < rows; i++ {
		rowSum := 0
		for j := 0; j < cols; j++ {
			rowSum += int(math.Abs(float64(matrix[i][j])))
		}
		if rowSum > maxNorm {
			maxNorm = rowSum
		}
	}
	return maxNorm
}

func MultiplyWithScalar(A, B [][]int, rows, cols int, scalar float64) {
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			B[i][j] = int(float64(A[i][j]) * scalar)
		}
	}
}

func Winograd(A [][]int, B [][]int) [][]int {

	N := len(B[0])
	P := len(B)
	M := len(A)

	Result := make([][]int, M)
	for i := range Result {
		Result[i] = make([]int, N)
	}

	var aux int
	upsilon := P % 2
	gamma := P - upsilon
	y := make([]int, M)
	z := make([]int, N)
	for i := 0; i < M; i++ {
		aux = 0
		for j := 0; j < gamma; j += 2 {
			aux += A[i][j] * A[i][j+1]
		}
		y[i] = aux
	}
	for i := 0; i < N; i++ {
		aux = 0
		for j := 0; j < gamma; j += 2 {
			aux += B[j][i] * B[j+1][i]
		}
		z[i] = aux
	}

	if upsilon == 1 {
		PP := P - 1
		for i := 0; i < M; i++ {
			for k := 0; k < N; k++ {
				aux = 0
				for j := 0; j < gamma; j += 2 {
					aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
				}
				Result[i][k] = aux - y[i] - z[k] + A[i][PP]*B[PP][k]
			}
		}
	} else {
		for i := 0; i < M; i++ {
			for k := 0; k < N; k++ {
				aux = 0
				for j := 0; j < gamma; j += 2 {
					aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
				}
				Result[i][k] = aux - y[i] - z[k]
			}
		}
	}

	return Result
}

File: python_algorithms/algorithms/WinogradScaled.py
import numpy as np
import math
import time

def Multiply(A, B):
    print("Método: WinogradScaled")

    start_time = time.time_ns()

    N = len(B[0])
    P = len(B)
    M = len(A)

    CopyA = [row.copy() for row in A]
    CopyB = [row.copy() for row in B]

    a = norm_inf(A, N, P)
    b = norm_inf(B, P, M)
    lambda_val = math.floor(0.5 + math.log(b / a) / math.log(4))

    multiply_with_scalar(A, CopyA, N, P, 2 ** lambda_val)
    multiply_with_scalar(B, CopyB, P, M, 2 ** -lambda_val)

    result = winograd(CopyA, CopyB)

    end_time = time.time_ns()
    execution_time = end_time - start_time

    return result, execution_time

def norm_inf(matrix, rows, cols):
    max_norm = 0
    for i in range(rows):
        row_sum = sum(abs(val) for val in matrix[i])
        max_norm = max(max_norm, row_sum)
    return max_norm

def multiply_with_scalar(A, B, rows, cols, scalar):
    for i in range(rows):
        for j in range(cols):
            B[i][j] = int(A[i][j] * scalar)

def winograd(A, B):
    N = len(B[0])
    P = len(B)
    M = len(A)

    Result = np.zeros((M, N), dtype=np.int64)

    upsilon = P % 2
    gamma = P - upsilon
    y = np.zeros(M, dtype=np.int64)
    z = np.zeros(N, dtype=np.int64)

    for i in range(M):
        aux = 0
        for j in range(0, gamma, 2):
            aux += A[i][j] * A[i][j+1]
        y[i] = aux

    for i in range(N):
        aux = 0
        for j in range(0, gamma, 2):
            aux += B[j][i] * B[j+1][i]
        z[i] = aux

    if upsilon == 1:
        PP = P - 1
        for i in range(M):
            for k in range(N):
                aux = 0
                for j in range(0, gamma, 2):
                    aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
                Result[i][k] = aux - y[i] - z[k] + A[i][PP] * B[PP][k]
    else:
        for i in range(M):
            for k in range(N):
                aux = 0
                for j in range(0, gamma, 2):
                    aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
                Result[i][k] = aux - y[i] - z[k]

    return Result
